apps/mobile/src/store/middleware/notifications.ts

```diff
@@ -0,0 +1,78 @@
+import type { Middleware, AnyAction } from '@reduxjs/toolkit'
+import type { RootState } from '..'
+import { addSafe, removeSafe, selectAllSafes } from '../safesSlice'
+import { subscribeSafe, unsubscribeSafe } from '@/src/services/notifications/SubscriptionManager'
+import { selectAllChainsIds } from '../chains'
+import { addDelegate } from '../delegatesSlice'
+import { selectSafeSubscriptionStatus } from '../safeSubscriptionsSlice'
+
+const notificationsMiddleware: Middleware = (store) => (next) => (action) => {
+  const typedAction = action as AnyAction
+  const prevState = store.getState() as RootState
+
+  const result = next(typedAction)
+
+  if (typedAction.type === addSafe.type) {
+    const { SafeInfo } = typedAction.payload
+    const notificationsEnabled = store.getState().notifications.isAppNotificationsEnabled
+    if (notificationsEnabled) {
+      const chainIds = selectAllChainsIds(store.getState())
+      subscribeSafe(SafeInfo.address.value, chainIds)
+    }
+  }
+
+  if (typedAction.type === removeSafe.type) {
+    const safeInfo = prevState.safes[typedAction.payload]
+    const chainIds = selectAllChainsIds(store.getState())
+    if (safeInfo) {
+      unsubscribeSafe(safeInfo.SafeInfo.address.value, chainIds)
+    }
+  }
+
+  if (typedAction.type === addDelegate.type) {
+    const { ownerAddress, delegateInfo } = typedAction.payload
+    const notificationsEnabled = store.getState().notifications.isAppNotificationsEnabled
+
+    if (notificationsEnabled) {
+      const chainIds = selectAllChainsIds(store.getState())
+      const safes = Object.values(selectAllSafes(store.getState()))
+      const state = store.getState()
+
+      safes.forEach((safe) => {
+        const safeAddress = safe.SafeInfo.address.value
+        const owners = safe.SafeInfo.owners.map((o) => o.value)
+        const isTargetSafe = delegateInfo.safe ? delegateInfo.safe === safeAddress : owners.includes(ownerAddress)
+
+        if (isTargetSafe) {
+          // Only subscribe if the Safe is already subscribed for notifications on at least one chain
+          const isSafeSubscribedOnAnyChain = chainIds.some(
+            (chainId) => selectSafeSubscriptionStatus(state, safeAddress, chainId) !== false,
+          )
+
+          if (isSafeSubscribedOnAnyChain) {
+            subscribeSafe(safeAddress, chainIds)
+          }
+        }
+      })
+    }
+  }
```

Comment:
In order to structure it better, we can try to follow a strategy pattern here. Where all the strategies implements the same interface and does their internal logic as we want. The benefit we could gain with this approach is a better separation of concern and it would be easier to cover the code with unit tests.